{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import xml.etree.ElementTree as et\n",
    "\n",
    "import numpy as np\n",
    "import scipy.sparse as sp\n",
    "import scipy.io as spio\n",
    "\n",
    "def reverse_complement(seq) :\n",
    "    seq_prime = ''\n",
    "    for j in range(0, len(seq)) :\n",
    "        if seq[j] == 'A' :\n",
    "            seq_prime = 'T' + seq_prime\n",
    "        elif seq[j] == 'C' :\n",
    "            seq_prime = 'G' + seq_prime\n",
    "        elif seq[j] == 'G' :\n",
    "            seq_prime = 'C' + seq_prime\n",
    "        elif seq[j] == 'T' :\n",
    "            seq_prime = 'A' + seq_prime\n",
    "\n",
    "    return seq_prime"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "apadb_bed    = 'hg19.apadb_v2_final.bed'#'../../data/genome_hg19/features/hg19.apadb_v2_final.bed'\n",
    "\n",
    "hg19_fai     = 'hg19.fa.fai'#'../../data/genome_hg19/hg19.fa.fai' \n",
    "hg19_fa      = 'hg19.fa'#'../../data/genome_hg19/hg19.fa'\n",
    "\n",
    "df_columns = ['chr', 'cut_start', 'cut_end', 'gene', 'reads', 'strand', 'feature', 'mode', 'miRNA']\n",
    "df = pd.read_csv(apadb_bed, sep='\\t', header=None, names=df_columns, usecols=[0,1,2,3,4,5,6,8,9])\n",
    "\n",
    "df = df.ix[df.miRNA == 'None']\n",
    "df = df.ix[df.cut_end - df.cut_start <= 90]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     chr  cut_start  cut_end       gene  reads strand feature    mode miRNA\n",
      "0  chr17      62260    62312   RPH3AL.1    122      -    UTR3   62297  None\n",
      "2  chr17     289758   289785  FAM101B.1     24      -    UTR3  289766  None\n",
      "3  chr17     289935   289947  FAM101B.2    170      -    UTR3  289946  None\n",
      "4  chr17     290016   290021  FAM101B.3     12      -    UTR3  290017  None\n",
      "5  chr17     290068   290076  FAM101B.4      6      -    UTR3  290071  None\n"
     ]
    }
   ],
   "source": [
    "print(df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     chr  cut_start  cut_end       gene  reads strand feature    mode miRNA  \\\n",
      "0  chr17      62260    62312   RPH3AL.1    122      -    UTR3   62297  None   \n",
      "2  chr17     289758   289785  FAM101B.1     24      -    UTR3  289766  None   \n",
      "3  chr17     289935   289947  FAM101B.2    170      -    UTR3  289946  None   \n",
      "4  chr17     290016   290021  FAM101B.3     12      -    UTR3  290017  None   \n",
      "5  chr17     290068   290076  FAM101B.4      6      -    UTR3  290071  None   \n",
      "\n",
      "    start     end  \n",
      "0   62222   62347  \n",
      "2  289691  289816  \n",
      "3  289871  289996  \n",
      "4  289942  290067  \n",
      "5  289996  290121  \n",
      "\n",
      "     chr  cut_start  cut_end       gene  reads strand feature    mode miRNA  \\\n",
      "0  chr17      62260    62312   RPH3AL.1    122      -    UTR3   62297  None   \n",
      "2  chr17     289758   289785  FAM101B.1     24      -    UTR3  289766  None   \n",
      "3  chr17     289935   289947  FAM101B.2    170      -    UTR3  289946  None   \n",
      "4  chr17     290016   290021  FAM101B.3     12      -    UTR3  290017  None   \n",
      "5  chr17     290068   290076  FAM101B.4      6      -    UTR3  290071  None   \n",
      "\n",
      "    start     end  genefam num_sites  total_reads  rel_use  \n",
      "0   62122   62472   RPH3AL         1          122    1.000  \n",
      "2  289591  289941  FAM101B         4          212    0.113  \n",
      "3  289771  290121  FAM101B         4          212    0.802  \n",
      "4  289842  290192  FAM101B         4          212    0.057  \n",
      "5  289896  290246  FAM101B         4          212    0.028  \n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "\n",
    "#SNP search range surrounding PAS (+- 50bps)\n",
    "df.ix[df.strand == '+', 'start'] = df['mode']-75#50#mode#cut_start\n",
    "df.ix[df.strand == '+', 'end'] = df['mode']+50#25#mode#cut_start\n",
    "\n",
    "df.ix[df.strand == '-', 'start'] = df['mode']-75#25#mode#cut_end\n",
    "df.ix[df.strand == '-', 'end'] = df['mode']+50#50#mode#cut_end\n",
    "\n",
    "df.start = df.start.astype(int)\n",
    "df.end = df.end.astype(int)\n",
    "\n",
    "print(df.head())\n",
    "print('')\n",
    "\n",
    "output_bed = 'apadb_shortsites.bed'\n",
    "bed_columns = ['chr', 'start', 'end', 'gene', 'reads', 'strand']\n",
    "df.to_csv(output_bed, sep='\\t', header=False, columns=bed_columns, index=False)\n",
    "\n",
    "\n",
    "\n",
    "#Create bed file of apadb sites\n",
    "df.ix[df.strand == '+', 'start'] = df['mode']-175\n",
    "df.ix[df.strand == '+', 'end'] = df['mode']+175\n",
    "\n",
    "df.ix[df.strand == '-', 'start'] = df['mode']-175\n",
    "df.ix[df.strand == '-', 'end'] = df['mode']+175\n",
    "\n",
    "df.start = df.start.astype(int)\n",
    "df.end = df.end.astype(int)\n",
    "\n",
    "df['genefam']     = df['gene'].str.split('\\\\.').apply(lambda x: x[0])\n",
    "df['num_sites']         = df.groupby('genefam')['genefam'].transform(len)\n",
    "df['total_reads'] = df.groupby('genefam')['reads'].transform(lambda x: sum(x))\n",
    "df['rel_use']     = (df['reads'] / df['total_reads']).round(3)\n",
    "\n",
    "print(df.head())\n",
    "print('')\n",
    "\n",
    "output_bed = 'apadb_sites.bed'\n",
    "bed_columns = ['chr', 'start', 'end', 'gene', 'reads', 'strand']\n",
    "df.to_csv(output_bed, sep='\\t', header=False, columns=bed_columns, index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr17  62122   62472   RPH3AL.1   122  -\n",
      "chr17  289591  289941  FAM101B.1  24   -\n",
      "chr17  289771  290121  FAM101B.2  170  -\n",
      "chr17  289842  290192  FAM101B.3  12   -\n",
      "chr17  289896  290246  FAM101B.4  6    -\n",
      "\n",
      ">RPH3AL.1\n",
      "CATCCTTCCTGCCTCCTGCCCCAAACCCCGGGTTCCTGGGTCTGGAAGGGCCTTCTCTCCAAGCTGGGAGCTCCTGGGCCCCCACCATTCACTTTTTGTCCTTGCTGCTGGCAAACAGTAAAGAAACTCACTTTCCCTGTGGCACGTTATGCTTCAGAATTAAAACAATGAAGATTAAAATTTGCACCGAGCCAGTGTGTTGATCGAAGACCACGATTGCCTGTGTTTCTGAGATGCGTCCATGGAAAAATGGAAAAAACTGTGGTGCGTTGACTTGCTGGAACCCTTCCTAAGCCGCAGTGAAAGGAGGGGCTAGATCTGTGTGTGTTCATGCAGCTCACACACCTGAT\n",
      ">FAM101B.1\n",
      "AAACTTAATTTGAGCGAGTACCTTTTCATTTGACACTTTTCCTGTTTCTAACCTTAGGAAACCAGAATAGCGTTTGGCAGACACGACGTTTTCAGTTTACCTTTGACACCTGCCCCACTCCATTTTGCTTTGTGATGTCTTCATTTAACAATAAATTATCTGAAAAAACAAAACTTAGAGAGATGCTTCTGTTTTTAAAGTAGAATTATGTTTGTTTACGCAAAATGAGAAAAACAGCTCCTCATCTTGAGAAATTTTAAGACGTGATTATATTTAACAGTATTAATCTACAAGTACAAGATTTTCCGAGTGTggctgggcatggtggctcacacctataatcccagcgc\n",
      ">FAM101B.2\n",
      "GGAAGCAGTGATTTTTAGGACCCACTGATTAAAAACAAACATTCCCAAGTGTCTCTGAGAGATGCTGTTTATTTGTTAATTAAAAAGCTTTTTTCTCTGTCTTTTAAATTATGGCTTTCATGTAATAAGGATATTTTTAGTGAAAAATTGTTTTCCTTTCAAATTACAGACCTTTTAAAAAAACTTAATTTGAGCGAGTACCTTTTCATTTGACACTTTTCCTGTTTCTAACCTTAGGAAACCAGAATAGCGTTTGGCAGACACGACGTTTTCAGTTTACCTTTGACACCTGCCCCACTCCATTTTGCTTTGTGATGTCTTCATTTAACAATAAATTATCTGAAAAAACA\n",
      ">FAM101B.3\n",
      "GGAGAGAAACAGAACCCTTCAGGGTGGGTCAGAGGACGCCATCCACAGTGGATTCGTGTTCGTTTGCAGGTGGAAGCAGTGATTTTTAGGACCCACTGATTAAAAACAAACATTCCCAAGTGTCTCTGAGAGATGCTGTTTATTTGTTAATTAAAAAGCTTTTTTCTCTGTCTTTTAAATTATGGCTTTCATGTAATAAGGATATTTTTAGTGAAAAATTGTTTTCCTTTCAAATTACAGACCTTTTAAAAAAACTTAATTTGAGCGAGTACCTTTTCATTTGACACTTTTCCTGTTTCTAACCTTAGGAAACCAGAATAGCGTTTGGCAGACACGACGTTTTCAGTTTA\n",
      ">FAM101B.4\n",
      "AATCCTCTTTCTTCCAAACCTGCTTTGCTGAAAGTTGCAGAAAAGGAAGTGTGTGGAGAGAAACAGAACCCTTCAGGGTGGGTCAGAGGACGCCATCCACAGTGGATTCGTGTTCGTTTGCAGGTGGAAGCAGTGATTTTTAGGACCCACTGATTAAAAACAAACATTCCCAAGTGTCTCTGAGAGATGCTGTTTATTTGTTAATTAAAAAGCTTTTTTCTCTGTCTTTTAAATTATGGCTTTCATGTAATAAGGATATTTTTAGTGAAAAATTGTTTTCCTTTCAAATTACAGACCTTTTAAAAAAACTTAATTTGAGCGAGTACCTTTTCATTTGACACTTTTCCTGT\n",
      "\n"
     ]
    }
   ],
   "source": [
    "hg19_fai     = 'hg19.fa.fai'\n",
    "hg19_fa      = 'hg19.fa'\n",
    "\n",
    "# fasta\n",
    "output_fa = 'apadb_seqs.fa'\n",
    "#!bedtools getfasta -name -s -fi \"$hg19_fa\" -bed \"$output_bed\" -fo | cut -d : -f-4 > \"$output_fa\"\n",
    "!bedtools getfasta -name -s -fi \"$hg19_fa\" -bed \"$output_bed\" -fo \"$output_fa\"\n",
    "    \n",
    "# file tops\n",
    "!head -5 \"$output_bed\" | column -t ; echo\n",
    "!head -10 \"$output_fa\" ; echo\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'total_reads': 212, 'chr': 'chr17', 'seq': 'GGAGAGAAACAGAACCCTTCAGGGTGGGTCAGAGGACGCCATCCACAGTGGATTCGTGTTCGTTTGCAGGTGGAAGCAGTGATTTTTAGGACCCACTGATTAAAAACAAACATTCCCAAGTGTCTCTGAGAGATGCTGTTTATTTGTTAATTAAAAAGCTTTTTTCTCTGTCTTTTAAATTATGGCTTTCATGTAATAAGGATATTTTTAGTGAAAAATTGTTTTCCTTTCAAATTACAGACCTTTTAAAAAAACTTAATTTGAGCGAGTACCTTTTCATTTGACACTTTTCCTGTTTCTAACCTTAGGAAACCAGAATAGCGTTTGGCAGACACGACGTTTTCAGTTTA', 'dn_site_dist': 71, 'start': 289842, 'mode': 290017, 'up_site_dist': 54, 'end': 290192, 'feature': 'UTR3', 'reads': 12, 'num_sites': 4, 'rel_use': 0.057, 'strand': '-'}\n"
     ]
    }
   ],
   "source": [
    "\n",
    "apadb_dict = {}\n",
    "\n",
    "for index, row in df.iterrows() :\n",
    "    apadb_dict[row['gene']] = {}\n",
    "    apadb_dict[row['gene']]['chr'] = row['chr']\n",
    "    apadb_dict[row['gene']]['start'] = row['start']\n",
    "    apadb_dict[row['gene']]['end'] = row['end']\n",
    "    apadb_dict[row['gene']]['mode'] = row['mode']\n",
    "    apadb_dict[row['gene']]['feature'] = row['feature']\n",
    "    apadb_dict[row['gene']]['strand'] = row['strand']\n",
    "    \n",
    "    apadb_dict[row['gene']]['num_sites'] = row['num_sites']\n",
    "    apadb_dict[row['gene']]['reads'] = row['reads']\n",
    "    apadb_dict[row['gene']]['total_reads'] = row['total_reads']\n",
    "    apadb_dict[row['gene']]['rel_use'] = row['rel_use']\n",
    "\n",
    "\n",
    "for gene in apadb_dict :\n",
    "    \n",
    "    genefam = gene.split('.')[0]\n",
    "    site = int(gene.split('.')[1])\n",
    "    \n",
    "    #Upstream site if any\n",
    "    up_gene = genefam + '.' + str(site + 1)\n",
    "    if up_gene in apadb_dict :\n",
    "        apadb_dict[gene]['up_site_dist'] = np.abs(apadb_dict[up_gene]['mode'] - apadb_dict[gene]['mode'])\n",
    "    else :\n",
    "        apadb_dict[gene]['up_site_dist'] = 10000\n",
    "    \n",
    "    #Downstream site if any\n",
    "    dn_gene = genefam + '.' + str(site - 1)\n",
    "    if dn_gene in apadb_dict :\n",
    "        apadb_dict[gene]['dn_site_dist'] = np.abs(apadb_dict[dn_gene]['mode'] - apadb_dict[gene]['mode'])\n",
    "    else :\n",
    "        apadb_dict[gene]['dn_site_dist'] = 10000\n",
    "\n",
    "\n",
    "with open('apadb_seqs.fa') as seq_f:\n",
    "    seq_id = ''\n",
    "    for line in seq_f:\n",
    "        if line[0] == '>' :\n",
    "            seq_id = line[1:len(line) - 1]\n",
    "        else :\n",
    "            apadb_dict[seq_id]['seq'] = line[:len(line) - 1].upper()\n",
    "            \n",
    "print(apadb_dict['FAM101B.3'])\n",
    "        \n",
    "        \n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "UTR-3\n",
      "STOP-GAIN\n",
      "129044\n",
      "{'cds-indel': True, 'intergenic': True, 'STOP-LOSS': True, 'UTR-3': True, 'splice-5': True, 'ncRNA': True, 'STOP-GAIN': True, 'intron': True, 'UTR-5': True, 'missense': True, 'splice-3': True, 'frameshift': True, 'cds-synon': True, 'nearGene-3': True, 'nearGene-5': True}\n"
     ]
    }
   ],
   "source": [
    "\n",
    "#Parse clinvar molecular conseq data\n",
    "\n",
    "\n",
    "cons_dict = {}\n",
    "\n",
    "unique_consequences = {}\n",
    "\n",
    "i = 0\n",
    "with open('molecular_consequences.txt') as f:\n",
    "    for line in f:\n",
    "        \n",
    "        lineparts = line[:-1].split('\\t')\n",
    "        \n",
    "        hgvs = lineparts[0]\n",
    "        cons = lineparts[2]\n",
    "        \n",
    "        if i > 0 :\n",
    "            cons_dict[hgvs] = cons\n",
    "            \n",
    "            if cons not in unique_consequences :\n",
    "                unique_consequences[cons] = True\n",
    "        \n",
    "        i += 1\n",
    "\n",
    "\n",
    "print(cons_dict['AJ132917.1:c.*14G>A'])\n",
    "\n",
    "print(cons_dict['NM_000047.2:c.1743G>A'])\n",
    "\n",
    "print(len(cons_dict))\n",
    "\n",
    "print(unique_consequences)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Parsed 0 rows.\n",
      "Parsed 10000000 rows.\n",
      "Parsed 20000000 rows.\n",
      "Parsed 30000000 rows.\n",
      "Parsed 40000000 rows.\n",
      "Parsed 50000000 rows.\n",
      "Parsed 60000000 rows.\n",
      "Parsed 70000000 rows.\n",
      "Parsed 80000000 rows.\n",
      "Parsed 90000000 rows.\n",
      "{'intron variant': True, '2KB upstream variant': True, 'splice acceptor variant': True, 'Splice Site donor': True, 'synonymous mutation': True, 'Read-through mutation': True, 'stop lost': True, 'splice donor variant': True, 'frameshift variant': True, 'synonymous variant': True, 'non-coding transcript variant': True, '500B downstream variant': True, 'nonsense': True, 'regulatory region ablation': True, 'missense variant': True, 'Splice Site acceptor': True, 'Silent': True, 'intergenic_variant': True, 'no-stop change?': True, 'Frameshift': True, 'frameshift mutation': True, '3 prime UTR variant': True, 'missense mutation': True, 'inframe_variant': True, 'Missense': True, 'Nonsense': True, '5 prime UTR variant': True, 'exon_loss': True}\n",
      "433704\n"
     ]
    }
   ],
   "source": [
    "\n",
    "#cons_dict = {}\n",
    "\n",
    "unique_consequences = {}\n",
    "\n",
    "i = 0\n",
    "with open('ClinVarFullRelease_2017-09.xml') as f:\n",
    "    \n",
    "    has_read_consequence = False\n",
    "    has_read_id = False\n",
    "    \n",
    "    consequence = ''\n",
    "    cons_id = ''\n",
    "    \n",
    "    for line in f:\n",
    "        \n",
    "        if i % 10000000 == 0 :\n",
    "            print('Parsed ' + str(i) + ' rows.')\n",
    "        \n",
    "        if 'MolecularConsequence' in line :\n",
    "            has_read_consequence = True\n",
    "            has_read_id = False\n",
    "            \n",
    "            consequence = line.split('<Attribute Type=\\\"MolecularConsequence\\\">')[1].split('</Attribute>')[0]\n",
    "            \n",
    "            if consequence not in unique_consequences :\n",
    "                unique_consequences[consequence] = True\n",
    "        \n",
    "        if has_read_consequence == True and has_read_id == False :\n",
    "            if 'XRef ID' in line and 'RefSeq' in line :\n",
    "                has_read_id = True\n",
    "                \n",
    "                cons_id = line.split('<XRef ID=\\\"')[1].split('\\\" DB=\"RefSeq\"/>')[0].replace('&gt;', '>')\n",
    "                \n",
    "                if cons_id not in cons_dict :\n",
    "                    cons_dict[cons_id] = consequence\n",
    "                \n",
    "                \n",
    "                consequence = ''\n",
    "                cons_id = ''\n",
    "        i += 1\n",
    "                \n",
    "print(unique_consequences)\n",
    "\n",
    "print(len(cons_dict))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "stop lost\n"
     ]
    }
   ],
   "source": [
    "print(cons_dict['NM_020461.3:c.5458T>G'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\n",
    "#Manually identified consequences\n",
    "cons_dict['NG_017041.1:g.5147G>A'] = '2KB upstream variant'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\n",
    "consequence_blacklist = []\n",
    "consequence_blacklist.append('cds-indel')\n",
    "consequence_blacklist.append('STOP-LOSS')\n",
    "consequence_blacklist.append('splice-5')\n",
    "consequence_blacklist.append('ncRNA')\n",
    "consequence_blacklist.append('STOP-GAIN')\n",
    "consequence_blacklist.append('intron')\n",
    "consequence_blacklist.append('missense')\n",
    "consequence_blacklist.append('splice-3')\n",
    "consequence_blacklist.append('frameshift')\n",
    "consequence_blacklist.append('cds-synon')\n",
    "consequence_blacklist.append('intron variant')\n",
    "consequence_blacklist.append('2KB upstream variant')\n",
    "consequence_blacklist.append('splice acceptor variant')\n",
    "consequence_blacklist.append('Splice Site donor')\n",
    "consequence_blacklist.append('Read-through mutation')\n",
    "consequence_blacklist.append('stop lost')\n",
    "consequence_blacklist.append('splice donor variant')\n",
    "consequence_blacklist.append('frameshift variant')\n",
    "consequence_blacklist.append('nonsense')\n",
    "#consequence_blacklist.append('')#regulatory region ablation\n",
    "consequence_blacklist.append('missense variant')\n",
    "consequence_blacklist.append('Splice Site acceptor')\n",
    "consequence_blacklist.append('no-stop change?')\n",
    "consequence_blacklist.append('Frameshift')\n",
    "consequence_blacklist.append('frameshift mutation')\n",
    "consequence_blacklist.append('missense mutation')\n",
    "consequence_blacklist.append('inframe_variant')\n",
    "consequence_blacklist.append('Missense')\n",
    "consequence_blacklist.append('Nonsense')\n",
    "consequence_blacklist.append('exon_loss')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     chr     start       end  \\\n",
      "0   chr7   4820844   4820847   \n",
      "1   chr7   4781213   4781216   \n",
      "2   chr7   4827366   4827379   \n",
      "3   chr7   4787735   4787748   \n",
      "4  chr15  85342440  85342440   \n",
      "\n",
      "                                                gene       refallele  \\\n",
      "0  AP5Z1:chr7:4820844:4820847:GGAT->TGCTGTAAACTGT...            GGAT   \n",
      "1  AP5Z1:chr7:4781213:4781216:GGAT->TGCTGTAAACTGT...            GGAT   \n",
      "2       AP5Z1:chr7:4827366:4827379:GGACCTGCCCTGCT->-  GGACCTGCCCTGCT   \n",
      "3       AP5Z1:chr7:4787735:4787748:GGACCTGCCCTGCT->-  GGACCTGCCCTGCT   \n",
      "4                ZNF592:chr15:85342440:85342440:G->A               G   \n",
      "\n",
      "                varallele assembly            significance  \\\n",
      "0  TGCTGTAAACTGTAACTGTAAA   GRCh37              Pathogenic   \n",
      "1  TGCTGTAAACTGTAACTGTAAA   GRCh38              Pathogenic   \n",
      "2                       -   GRCh37              Pathogenic   \n",
      "3                       -   GRCh38              Pathogenic   \n",
      "4                       A   GRCh37  Uncertain significance   \n",
      "\n",
      "                     vartype         consequence  \\\n",
      "0                      indel     inframe_variant   \n",
      "1                      indel     inframe_variant   \n",
      "2                   deletion  frameshift variant   \n",
      "3                   deletion  frameshift variant   \n",
      "4  single nucleotide variant            missense   \n",
      "\n",
      "                                                  id  \n",
      "0  NM_014855.2(AP5Z1):c.80_83delGGATinsTGCTGTAAAC...  \n",
      "1  NM_014855.2(AP5Z1):c.80_83delGGATinsTGCTGTAAAC...  \n",
      "2  NM_014855.2(AP5Z1):c.1413_1426delGGACCTGCCCTGC...  \n",
      "3  NM_014855.2(AP5Z1):c.1413_1426delGGACCTGCCCTGC...  \n",
      "4       NM_014630.2(ZNF592):c.3136G>A (p.Gly1046Arg)  \n",
      "\n",
      "       chr      start        end                                      gene  \\\n",
      "86    chr2   63720075   63720075         WDPCP:chr2:63720075:63720075:C->A   \n",
      "88    chr2   63664564   63664564         WDPCP:chr2:63664564:63664564:C->G   \n",
      "101  chr12  123738469  123738469   C12orf65:chr12:123738469:123738469:T->-   \n",
      "200   chr9  130548229  130548229     MIR2861:chr9:130548229:130548229:C->G   \n",
      "211  chr14   77907418   77907422  VIPAS39:chr14:77907418:77907422:TTCTG->-   \n",
      "\n",
      "    refallele varallele assembly                  significance  \\\n",
      "86          C         A   GRCh37                    Pathogenic   \n",
      "88          C         G   GRCh37                   risk factor   \n",
      "101         T         -   GRCh37  Pathogenic/Likely pathogenic   \n",
      "200         C         G   GRCh37                   association   \n",
      "211     TTCTG         -   GRCh37                    Pathogenic   \n",
      "\n",
      "                       vartype                    consequence  \\\n",
      "86   single nucleotide variant                   undetermined   \n",
      "88   single nucleotide variant                   undetermined   \n",
      "101                   deletion                   undetermined   \n",
      "200  single nucleotide variant  non-coding transcript variant   \n",
      "211                   deletion                   undetermined   \n",
      "\n",
      "                                                    id  \n",
      "86                        NM_015910.6(WDPCP):c.76-1G>T  \n",
      "88           NM_015910.6(WDPCP):c.624G>C (p.Leu208Phe)  \n",
      "101  NM_001143905.2(C12orf65):c.248delT (p.Val83Glyfs)  \n",
      "200                       NR_036055.1(MIR2861):n.33C>G  \n",
      "211  NM_022067.3(VIPAS39):c.749_753delCAGAA (p.Thr2...  \n",
      "110195\n"
     ]
    }
   ],
   "source": [
    "\n",
    "#Parse clinvar data\n",
    "clinvar_summary_file = 'variant_summary.txt'\n",
    "\n",
    "variant_type = []\n",
    "variant_gene = []\n",
    "variant_clinsig = []\n",
    "variant_assembly = []\n",
    "variant_chrom = []\n",
    "variant_start = []\n",
    "variant_end = []\n",
    "variant_refallele = []\n",
    "variant_varallele = []\n",
    "variant_key = []\n",
    "variant_id = []\n",
    "\n",
    "variant_cons = []\n",
    "\n",
    "i = 0\n",
    "with open(clinvar_summary_file) as f:\n",
    "    for line in f:\n",
    "        \n",
    "        lineparts = line.split('\\t')\n",
    "        \n",
    "        if i > 0 :\n",
    "            variant_type.append(lineparts[1])\n",
    "            variant_gene.append(lineparts[4] + ':' + 'chr' + lineparts[18] + ':' + lineparts[19] + ':' + lineparts[20] + ':' + lineparts[21] + '->' + lineparts[22])\n",
    "            variant_clinsig.append(lineparts[6])\n",
    "            variant_assembly.append(lineparts[16])\n",
    "            variant_chrom.append('chr' + lineparts[18])\n",
    "            variant_start.append(int(lineparts[19]))\n",
    "            variant_end.append(int(lineparts[20]))\n",
    "            variant_refallele.append(lineparts[21])\n",
    "            variant_varallele.append(lineparts[22])\n",
    "            \n",
    "            variant_id.append(lineparts[2])\n",
    "            \n",
    "            hgvs = ''\n",
    "            \n",
    "            hgvs = lineparts[2]\n",
    "            if ' ' in hgvs :\n",
    "                hgvs = hgvs.split(' ')[0]\n",
    "            \n",
    "            if len(hgvs.split('(')) == 2 :\n",
    "                hgvs = hgvs.split('(')[0] + hgvs.split(')')[1]\n",
    "            \n",
    "            \n",
    "            if hgvs in cons_dict:\n",
    "                variant_cons.append(cons_dict[hgvs])\n",
    "            else :\n",
    "                variant_cons.append('undetermined')\n",
    "        \n",
    "        i += 1\n",
    "\n",
    "clinvar_df = pd.DataFrame({'chr' : variant_chrom,\n",
    "                    'start' : variant_start,\n",
    "                    'end' : variant_end,\n",
    "                    'gene' : variant_gene,\n",
    "                    'refallele' : variant_refallele,\n",
    "                    'varallele' : variant_varallele,\n",
    "                    'assembly' : variant_assembly,\n",
    "                    'significance' : variant_clinsig,\n",
    "                    'vartype' : variant_type,\n",
    "                    'consequence' : variant_cons,\n",
    "                    'id' : variant_id\n",
    "                })\n",
    "\n",
    "clinvar_df = clinvar_df[['chr', 'start', 'end', 'gene', 'refallele', 'varallele', 'assembly', 'significance', 'vartype', 'consequence', 'id']]\n",
    "\n",
    "\n",
    "print(clinvar_df.head())\n",
    "print('')\n",
    "\n",
    "clinvar_df = clinvar_df.ix[clinvar_df.assembly == 'GRCh37']\n",
    "#clinvar_df = clinvar_df.ix[clinvar_df.vartype == 'single nucleotide variant']\n",
    "clinvar_df = clinvar_df.ix[(((clinvar_df.vartype == 'insertion') | (clinvar_df.vartype == 'deletion')) | (clinvar_df.vartype == 'indel')) | (clinvar_df.vartype == 'single nucleotide variant')]\n",
    "clinvar_df = clinvar_df.ix[clinvar_df.end - clinvar_df.start <= 20]\n",
    "\n",
    "'''\n",
    "clinvar_df = clinvar_df.ix[\n",
    "    (((((((clinvar_df.consequence == 'UTR-3')\n",
    "    |\n",
    "    (clinvar_df.consequence == 'UTR-51'))\n",
    "    |\n",
    "    (clinvar_df.consequence == 'intron1'))\n",
    "    |\n",
    "    (clinvar_df.consequence == 'undetermined'))\n",
    "    |\n",
    "    (clinvar_df.consequence == '5 prime UTR variant1'))\n",
    "    |\n",
    "    (clinvar_df.consequence == '3 prime UTR variant'))\n",
    "    |\n",
    "    (clinvar_df.consequence == 'non-coding transcript variant'))\n",
    "    |\n",
    "    (clinvar_df.consequence == 'intron variant1')\n",
    "]'''\n",
    "\n",
    "for conseq_blacklisted in consequence_blacklist :\n",
    "    clinvar_df = clinvar_df.ix[clinvar_df.consequence != conseq_blacklisted]\n",
    "\n",
    "print(clinvar_df.head())\n",
    "print(len(clinvar_df))\n",
    "#59422\n",
    "\n",
    "\n",
    "clinvar_dict = {}\n",
    "\n",
    "for index, row in clinvar_df.iterrows() :\n",
    "    clinvar_dict[row['gene']] = {}\n",
    "    clinvar_dict[row['gene']]['significance'] = row['significance']\n",
    "    clinvar_dict[row['gene']]['vartype'] = row['vartype']\n",
    "    clinvar_dict[row['gene']]['varcons'] = row['consequence']\n",
    "    \n",
    "    clinvar_dict[row['gene']]['refallele'] = row['refallele']\n",
    "    clinvar_dict[row['gene']]['varallele'] = row['varallele']\n",
    "    \n",
    "    clinvar_dict[row['gene']]['id'] = row['id']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Create bed file of apadb sites\n",
    "\n",
    "clinvar_bed = 'clinvar_snps.bed'\n",
    "bed_columns = ['chr', 'start', 'end', 'gene']\n",
    "clinvar_df.to_csv(clinvar_bed, sep='\\t', header=False, columns=bed_columns, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\n",
    "#Intersect Clinvar snps against APADB bed\n",
    "\n",
    "!bedtools intersect -wa -wb -a 'clinvar_snps.bed' -b 'apadb_shortsites.bed' > 'clinvar_apadb_intersection.bed'\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     chr  snp_start    snp_end                               snp_key  \\\n",
      "0  chr22   45691707   45691707    UPK3A:chr22:45691707:45691707:T->C   \n",
      "1   chr7  128589427  128589427    IRF5:chr7:128589427:128589427:G->A   \n",
      "2  chr12  102122887  102122887  SYCP3:chr12:102122887:102122887:A->G   \n",
      "3   chr2  231050715  231050715   SP110:chr2:231050715:231050715:A->G   \n",
      "4   chr1   24020362   24020362     RPL11:chr1:24020362:24020362:C->T   \n",
      "\n",
      "  apadb_site  \n",
      "0    UPK3A.1  \n",
      "1     IRF5.2  \n",
      "2    CHPT1.3  \n",
      "3   SP110.10  \n",
      "4    RPL11.4  \n",
      "3483\n"
     ]
    }
   ],
   "source": [
    "\n",
    "intersect_columns = ['chr', 'snp_start', 'snp_end', 'snp_key', 'apadb_site']\n",
    "intersect_df = pd.read_csv('clinvar_apadb_intersection.bed', sep='\\t', header=None, names=intersect_columns, usecols=[0, 1, 2, 3, 7])\n",
    "\n",
    "print(intersect_df.head())\n",
    "\n",
    "print(len(intersect_df))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Empty DataFrame\n",
      "Columns: [chr, snp_start, snp_end, snp_key, apadb_site]\n",
      "Index: []\n"
     ]
    }
   ],
   "source": [
    "print(intersect_df[intersect_df.apadb_site == 'TUBGCP6.2'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING! Ref seq == Var seq.\n",
      "WARNING (-)! Ref base differs.\n",
      "indel\n",
      "206\n",
      "TGTGCGCGCGCGCACGCGCGAGTGTGCTGTATGGCCCAGGCAGCCTCAAGGCCCTCGGAGCTGGCTGTGCCTGCTTCTGTGTACCACTTCTGTGGGCATGGCCGCTTCTAGAGCCTCGACACCCCCCCAACCCCCGCACCAAGCAGACAAAGTCAATAAAAGAGCTGTCTGACTGCAATCTGTGCCTCTATGTCTGTGCACTGGGGTCAGGACTTTATTTATTTCACTGACAGGCAATACCGTCCAAGGCCAGTGCAGGAGGGAGGGCCCCGGCCTCACACAAACTCGGTGAAGTCCTCCACCGAGGAGATGAGGCGCTTCCGCTGGCCCACCTCATAGCCAGGTGTGGG\n",
      "TGCACTGGGGTCAGGACTTTATT\n",
      "TCA\n",
      "\n",
      "WARNING (+)! Ref base differs.\n",
      "indel\n",
      "147\n",
      "AGTCCTGACCCCA\n",
      "TGA\n",
      "na\n",
      "WARNING (-)! Ref base differs.\n",
      "deletion\n",
      "176\n",
      "CGCACAGCAACATTTTGAGAATAATCTTCTACTAATGAATGGTAGTGAGTTGAAAATAATGTACGACATTTTATAGTCTCAGCAAGTTCTTTAACAACTGCATTTGCTATTGCCGTCCCATCAAATGTTGCAGTACCTCTTCCTGTTAAAGTAAAATATGCATAAGGAAGTAACTCAAAGGAATTAAAACAAAAAAGGAATTAAAACAAAAATGCTAGGACAGAAAAGCAACATCGGTTAGTACATCCACGTCTAAAAGCATTCTATAAATAGGCCTTGTTTAGCTACACGAGTCTGCTTAGCAGCCCATGGGGAGTGAGGTTTCTTTTAAAGAAAATACAGTATCGCTC\n",
      "GAAGTAACTCAAAGGAATTAAAACAAAAAAGGAATTA\n",
      "AAAGGAATTAAAACAAA\n",
      "\n",
      "WARNING (-)! Ref base differs.\n",
      "indel\n",
      "249\n",
      "AAAATATTTGTCAAAATCTTAACTGAATGTTTACTGGAAGTACTTGAGATTCCATTTGAGAGTTGTATTGTTAATAATTTCATGTCAGTGAACTGATATCTGATGTTTATGATATGGTGTCTTTTTCTTGAAACAAGCTTCCAAGGGCTAGAAATAAAATAGCCAAAAAATGCTGGAGTTCTGAGTAATTAAATGGCAGCATTTTTTTGTGACAAAGGTAAGGGTTAAGACAGTGTGTGTGTGTGTGTGTGTGTGTGTGTGTGTGTGTGTGCGCGCGCGCGCGCGCGCGTAACGTGGGCGCATCCCTCTTTTCCTTCCCTGGGTTCAGTGTGGCTGGGTAAGGACCGTCT\n",
      "TGTGTGTGTGTGTGTGTGTGTGTGTGTGTGTGCGCGCGCGC\n",
      "TGTGTGTGTGTGTGTGTGTGT\n",
      "\n",
      "WARNING! Ref seq == Var seq.\n",
      "WARNING (-)! Ref base differs.\n",
      "indel\n",
      "249\n",
      "AAAATATTTGTCAAAATCTTAACTGAATGTTTACTGGAAGTACTTGAGATTCCATTTGAGAGTTGTATTGTTAATAATTTCATGTCAGTGAACTGATATCTGATGTTTATGATATGGTGTCTTTTTCTTGAAACAAGCTTCCAAGGGCTAGAAATAAAATAGCCAAAAAATGCTGGAGTTCTGAGTAATTAAATGGCAGCATTTTTTTGTGACAAAGGTAAGGGTTAAGACAGTGTGTGTGTGTGTGTGTGTGTGTGTGTGTGTGTGTGTGCGCGCGCGCGCGCGCGCGTAACGTGGGCGCATCCCTCTTTTCCTTCCCTGGGTTCAGTGTGGCTGGGTAAGGACCGTCT\n",
      "TGTGTGTGTGTGTGTGTGTGTGTGTGTGTGTGCGCGCGCGC\n",
      "TGTGTGTGTGTGTGTGTGTGT\n",
      "\n",
      "         gene  var_pos                                            ref_seq  \\\n",
      "2535   AAAS.1      179  TGGGCGGGCCCAGGAACCCCCTGCTGGGGGTGGAGGCTCTATTCAT...   \n",
      "2969   AARS.2      165  AATTTCTGCCCTGAGCCCTCCACGTCAGTGCCATCGGTCTAGAACC...   \n",
      "1487  AARS2.1      171  TGGAGGAAAGCCCTCCTTCCCTACCAGAATGAGTCCACCTCAGACA...   \n",
      "1489  AARS2.2      146  GTGAGGGTGTGACCCGTGTGGAAATCAGGGAAAAGCATTCCAGCCT...   \n",
      "1488  AARS2.2      232  GTGAGGGTGTGACCCGTGTGGAAATCAGGGAAAAGCATTCCAGCCT...   \n",
      "\n",
      "                                                var_seq region  \\\n",
      "2535  TGGGCGGGCCCAGGAACCCCCTGCTGGGGGTGGAGGCTCTATTCAT...   UTR3   \n",
      "2969  AATTTCTGCCCTGAGCCCTCCACGTCAGTGCCATCGGTCTAGAACC...   UTR3   \n",
      "1487  TGGAGGAAAGCCCTCCTTCCCTACCAGAATGAGTCCACCTCAGACA...   UTR3   \n",
      "1489  GTGAGGGTGTGACCCGTGTGGAAATCAGGGAAAAGCATTCCAGCCT...   UTR3   \n",
      "1488  GTGAGGGTGTGACCCGTGTGGAAATCAGGGAAAAGCATTCCAGCCT...   UTR3   \n",
      "\n",
      "                significance                    vartype  num_sites  \\\n",
      "2535           Likely benign  single nucleotide variant          1   \n",
      "2969  Uncertain significance  single nucleotide variant          2   \n",
      "1487  Uncertain significance  single nucleotide variant          5   \n",
      "1489  Uncertain significance  single nucleotide variant          5   \n",
      "1488                  Benign                  insertion          5   \n",
      "\n",
      "      up_site_dist  dn_site_dist  reads  total_reads  rel_use  \\\n",
      "2535         10000         10000   1090         1090    1.000   \n",
      "2969         10000            89   1131         1814    0.623   \n",
      "1487           237         10000    114          297    0.384   \n",
      "1489           592           237     20          297    0.067   \n",
      "1488           592           237     20          297    0.067   \n",
      "\n",
      "              consequence                             clinvar_id  \n",
      "2535  3 prime UTR variant             NM_015665.5(AAAS):c.*32C>T  \n",
      "2969  3 prime UTR variant            NM_001605.2(AARS):c.*324G>A  \n",
      "1487  3 prime UTR variant          NM_020745.3(AARS2):c.*1811C>T  \n",
      "1489  3 prime UTR variant          NM_020745.3(AARS2):c.*1549C>T  \n",
      "1488  3 prime UTR variant  NM_020745.3(AARS2):c.*1635_*1636insAA  \n",
      "3476\n"
     ]
    }
   ],
   "source": [
    "\n",
    "def reverse_complement(seq) :\n",
    "    rev_seq = seq[::-1]\n",
    "    rev_comp = ''\n",
    "    for base in rev_seq :\n",
    "        if base == 'A' :\n",
    "            rev_comp += 'T'\n",
    "        elif base == 'C' :\n",
    "            rev_comp += 'G'\n",
    "        elif base == 'G' :\n",
    "            rev_comp += 'C'\n",
    "        elif base == 'T' :\n",
    "            rev_comp += 'A'\n",
    "    return rev_comp\n",
    "\n",
    "\n",
    "#Generate APADB Clinvar variant dataset\n",
    "\n",
    "l_gene = []\n",
    "l_refseq = []\n",
    "l_varseq = []\n",
    "l_region = []\n",
    "l_significance = []\n",
    "l_vartype = []\n",
    "l_varpos = []\n",
    "l_reads = []\n",
    "l_totalreads = []\n",
    "l_reluse = []\n",
    "l_numsites = []\n",
    "l_upsitedist = []\n",
    "l_dnsitedist = []\n",
    "l_consequence = []\n",
    "l_id = []\n",
    "\n",
    "for index, row in intersect_df.iterrows() :\n",
    "    site_id = row['apadb_site']\n",
    "    \n",
    "    site_start = apadb_dict[site_id]['start']\n",
    "    site_end = apadb_dict[site_id]['end']\n",
    "    site_strand = apadb_dict[site_id]['strand']\n",
    "    site_refseq = apadb_dict[site_id]['seq']\n",
    "    \n",
    "    site_num = apadb_dict[site_id]['num_sites']\n",
    "    site_up_dist = apadb_dict[site_id]['up_site_dist']\n",
    "    site_dn_dist = apadb_dict[site_id]['dn_site_dist']\n",
    "    \n",
    "    snp_start = row['snp_start']\n",
    "    snp_end = row['snp_end']\n",
    "    #snp_ref = row['snp_key'].split('->')[0][len(row['snp_key'].split('->')[0])-1]\n",
    "    #snp_var = row['snp_key'].split('->')[1][0]\n",
    "    \n",
    "    snp_ref = clinvar_dict[row['snp_key']]['refallele']\n",
    "    snp_var = clinvar_dict[row['snp_key']]['varallele']\n",
    "    \n",
    "    snp_type = clinvar_dict[row['snp_key']]['vartype']\n",
    "    \n",
    "    snp_cons = clinvar_dict[row['snp_key']]['varcons']\n",
    "    \n",
    "    snp_id = clinvar_dict[row['snp_key']]['id']\n",
    "    \n",
    "    if snp_ref == '-' :\n",
    "        snp_ref = ''\n",
    "    if snp_var == '-' :\n",
    "        snp_var = ''\n",
    "    \n",
    "    \n",
    "    site_varseq = site_refseq\n",
    "    relpos_start = -1\n",
    "    relpos_end = -1\n",
    "    \n",
    "    if site_strand == '+' :\n",
    "        relpos_start = snp_start - site_start - 1\n",
    "        relpos_end = snp_start - site_start - 1 + (snp_end - snp_start)\n",
    "        \n",
    "        if relpos_start < 0 :\n",
    "            print('WARNING (+)! Relpos out of range (' + str(relpos_start) + ')')\n",
    "            continue\n",
    "        if site_refseq[relpos_start:relpos_end+1] != snp_ref and (snp_type == 'deletion' or snp_type == 'indel'):\n",
    "            print('WARNING (+)! Ref base differs.')\n",
    "            print(snp_type)\n",
    "            print(relpos_start)\n",
    "            print(site_refseq[relpos_start-5:relpos_end+1+5])\n",
    "            print(site_refseq[relpos_start:relpos_end+1])\n",
    "            print(snp_ref)\n",
    "            continue\n",
    "            \n",
    "        site_varseq = site_varseq[:relpos_start] + snp_var + site_varseq[relpos_end+1:]\n",
    "    if site_strand == '-' :\n",
    "        \n",
    "        snp_ref = reverse_complement(snp_ref)\n",
    "        snp_var = reverse_complement(snp_var)\n",
    "        \n",
    "        relpos_end = 350 - (snp_start - site_start)\n",
    "        relpos_start = 350 - (snp_start - site_start + (snp_end - snp_start))\n",
    "        \n",
    "        if relpos_start < 0 :\n",
    "            print('WARNING (-)! Relpos out of range (' + str(relpos_start) + ')')\n",
    "            continue\n",
    "        if site_refseq[relpos_start:relpos_end+1] != snp_ref and (snp_type == 'deletion' or snp_type == 'indel'):\n",
    "            print('WARNING (-)! Ref base differs.')\n",
    "            print(snp_type)\n",
    "            print(relpos_start)\n",
    "            print(site_refseq)\n",
    "            print(site_refseq[relpos_start-10:relpos_end+1+10])\n",
    "            print(site_refseq[relpos_start:relpos_end+1])\n",
    "            print(snp_ref)\n",
    "            continue\n",
    "            \n",
    "        site_varseq = site_varseq[:relpos_start] + snp_var + site_varseq[relpos_end+1:]\n",
    "        \n",
    "    \n",
    "    if relpos_start < 150 and snp_type == 'deletion' :\n",
    "        site_varseq = ('X' * (350 - len(site_varseq))) + site_varseq\n",
    "    elif relpos_start < 150 and snp_type == 'insertion' :\n",
    "        site_varseq = site_varseq + ('X' * (350 - len(site_varseq)))\n",
    "    elif relpos_start < 150 and snp_type == 'indel' and len(snp_ref) > len(snp_var) :\n",
    "        site_varseq = ('X' * (350 - len(site_varseq))) + site_varseq\n",
    "    elif relpos_start < 150 and snp_type == 'indel' and len(snp_ref) < len(snp_var) :\n",
    "        site_varseq = site_varseq + ('X' * (350 - len(site_varseq)))\n",
    "    elif relpos_start >= 150 :\n",
    "        site_varseq = site_varseq + ('X' * (350 - len(site_varseq)))\n",
    "    else :\n",
    "        site_varseq = site_varseq + ('X' * (350 - len(site_varseq)))\n",
    "    \n",
    "    if site_refseq == site_varseq :\n",
    "        print('WARNING! Ref seq == Var seq.')\n",
    "        continue\n",
    "    \n",
    "    l_gene.append(site_id)\n",
    "    l_varpos.append(relpos_start)\n",
    "    l_refseq.append(site_refseq)\n",
    "    l_varseq.append(site_varseq)\n",
    "    l_region.append(apadb_dict[site_id]['feature'])\n",
    "    l_significance.append(clinvar_dict[row['snp_key']]['significance'])\n",
    "    l_vartype.append(clinvar_dict[row['snp_key']]['vartype'])\n",
    "        \n",
    "    l_reads.append(apadb_dict[site_id]['reads'])\n",
    "    l_totalreads.append(apadb_dict[site_id]['total_reads'])\n",
    "    l_reluse.append(apadb_dict[site_id]['rel_use'])\n",
    "    \n",
    "    l_numsites.append(site_num)\n",
    "    l_upsitedist.append(site_up_dist)\n",
    "    l_dnsitedist.append(site_dn_dist)\n",
    "    \n",
    "    l_consequence.append(snp_cons)\n",
    "    \n",
    "    l_id.append(snp_id)\n",
    "        \n",
    "        \n",
    "snp_df = pd.DataFrame({'gene' : l_gene,\n",
    "                    'var_pos' : l_varpos,\n",
    "                    'ref_seq' : l_refseq,\n",
    "                    'var_seq' : l_varseq,\n",
    "                    'region' : l_region,\n",
    "                    'significance' : l_significance,\n",
    "                    'vartype' : l_vartype,\n",
    "                    'num_sites' : l_numsites,\n",
    "                    'up_site_dist' : l_upsitedist,\n",
    "                    'dn_site_dist' : l_dnsitedist,\n",
    "                    'reads' : l_reads,\n",
    "                    'total_reads' : l_totalreads,\n",
    "                    'rel_use' : l_reluse,\n",
    "                    'consequence' : l_consequence,\n",
    "                    'clinvar_id' : l_id\n",
    "                })\n",
    "\n",
    "snp_df = snp_df[['gene', 'var_pos', 'ref_seq', 'var_seq', 'region', 'significance', 'vartype', 'num_sites', 'up_site_dist', 'dn_site_dist', 'reads', 'total_reads', 'rel_use', 'consequence', 'clinvar_id']]\n",
    "snp_df = snp_df.sort_values(by='gene')\n",
    "\n",
    "print(snp_df.head())\n",
    "print(len(snp_df))\n",
    "\n",
    "snp_df.to_csv('apadb_snps_combined_blacklist.csv', header=True, index=False, sep='\\t')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
